pkg/services/forwarder/ports.go

```diff
@@ -50,22 +55,165 @@ func (f *PortsForwarder) Expose(protocol types.TransportProtocol, local, remote
+		}
+
+		// build the address from remoteUri
+		remoteAddr := fmt.Sprintf("%s:%s", remoteUri.Hostname(), remoteUri.Port())
+
+		// dialFn opens remote connection for the proxy
+		var dialFn func(ctx context.Context, network, addr string) (conn net.Conn, e error)
+
+		// dialFn is set based on the protocol provided by remoteUri.Scheme
+		switch remoteUri.Scheme {
+		case "ssh-tunnel": // unix-to-unix proxy (over SSH)
+			// query string to map for the remoteUri contains ssh config info
+			remoteQuery := remoteUri.Query()
+
+			// username
+			sshuser := firstValueOrEmpty(remoteQuery["user"])
+			if sshuser == "" {
+				return fmt.Errorf("user not provided for unix-ssh connection")
+			}
+
+			// key
+			sshkeypath := firstValueOrEmpty(remoteQuery["key"])
+			if sshkeypath == "" {
+				return fmt.Errorf("key not provided for unix-ssh connection")
+			}
+
+			sshkeyBytes, err := ioutil.ReadFile(sshkeypath)
+			if err != nil {
+				return fmt.Errorf("failed to read ssh key: %s: %w", sshkeypath, err)
+			}
+
+			// passphrase
+			passphrase := firstValueOrEmpty(remoteQuery["passphrase"])
+
+			var sshsigner ssh.Signer
+
+			if passphrase == "" {
+				sshsigner, err = ssh.ParsePrivateKey(sshkeyBytes)
+			} else {
+				sshsigner, err = ssh.ParsePrivateKeyWithPassphrase(sshkeyBytes, []byte(passphrase))
+			}
+
+			// parse private key error?
+			if err != nil {
+				return fmt.Errorf("failed to parse ssh key: %s: %w", sshkeypath, err)
+			}
+
+			// default ssh port if not set
+			if remoteUri.Port() == "" {
+				remoteAddr = fmt.Sprintf("%s:%s", remoteUri.Hostname(), "22")
+			}
+
+			// build address
+			address, err := tcpipAddress(1, remoteAddr)
+			if err != nil {
+				return err
+			}
+
+			// check the remoteUri path provided for nonsense
+			if remoteUri.Path == "" || remoteUri.Path == "/" {
+				return fmt.Errorf("remote uri must contain a path to a socket file")
+			}
+
+			// the dialFn for unix-to-unix over SSH
+			dialFn = func(ctx context.Context, network, addr string) (conn net.Conn, e error) {
+				// underlying connection to endpoint for the ssh client
+				conn, err := gonet.DialContextTCP(ctx, f.stack, address, ipv4.ProtocolNumber)
+				if err != nil {
+					return nil, err
+				}
+
+				// ssh client config that uses key authentication
+				config := &ssh.ClientConfig{
+					User: sshuser,
+					Auth: []ssh.AuthMethod{
+						ssh.PublicKeys(sshsigner),
+					},
+					HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+					HostKeyAlgorithms: []string{
+						ssh.KeyAlgoRSA,
+						ssh.KeyAlgoDSA,
+						ssh.KeyAlgoECDSA256,
+						ssh.KeyAlgoECDSA384,
+						ssh.KeyAlgoECDSA521,
+						ssh.KeyAlgoED25519,
+					},
+					Timeout: 5 * time.Second,
+				}
+
+				// get an sshConn using the underlying gonet.TCPConn
+				sshConn, chans, reqs, err := ssh.NewClientConn(conn, addr, config)
+				if err != nil {
+					return nil, err
+				}
+
+				// build an ssh client using sshConn
+				sshClient := ssh.NewClient(sshConn, chans, reqs)
```

Comment:
First off thank you for contributing! I have some comments, but keep in mind I am just a contributor like you and not an official reviewer/merger 

It looks like this approach is creating a new ssh connection per client. One of the benefits of the existing implementation is that, just like OpenSSH's forwarding behavior, the ssh connection is kept alive and reused for multiple connections on the forwarded unix socket. This allows us to avoid paying key exchange costs with every client connect, fitting expectations of existing docker api clients coded against a unix socket. 

On this topic, in addition to your proposed dynamic registration addition, we have plans to add a ssh-forwarding only variant of gvproxy for windows that does not involve vsock (I hope to take a look at this sometime this week). Relatedly we've previously discussed it being a good idea to share the same ssh client code in the long term between podman and gvproxy to ensure consistent behavior and share bug fixes etc).  Ideally, everything uses the same code. Can this be updated to reuse the existing static impl (then you get stuff like the above for free)? 

As a general note, IMO the static preconfigured ssh forwarding is desirable in the podman machine use-case (aside from the non-vsock windows need) since it can be established immediately without blocking on subsequent round trips (similar to the preconfigured routes capability).

Reply:
Thanks for the patch @protosam and the review @n1hility!
I think both commits are useful. Can you add 2 end-to-end tests for them? You can find some examples in `test/`.

> It looks like this approach is creating a new ssh connection per client. 

Yes we can evolve it in the future and have a single or a pool of connections? 

> As a general note, IMO the static preconfigured ssh forwarding is desirable in the podman machine use-case (aside from the non-vsock windows need) since it can be established immediately without blocking on subsequent round trips (similar to the preconfigured routes capability).

`gvproxy` is really for podman right now, so we optimize for it. If you have other uses of this binary, @protosam, perhaps we can change it to have a configuration file or create a more generic binary?

I am happy to merge anything that can help podman machine (or crc).

Reply:
Will update this further after I figure out how the tests should work. I'm working on fixing the multiple SSH connections behavior and the tests this week when I have some time.

Thanks for the code review @n1hility

I'm going to patch it a single ssh connection is used per forwarder in this PR. The behavior as noted is actually a bug, the connection should be scoped to the forwarder expose function and dial function should be only opening a connection if it detects it was closed.

Reusing the existing ssh forwarder implementation looks like it adds unnecessary overhead for an implementation that lives as a networking service this close to the networking stack. (Disclaimer, I might not know what I'm even talking about here, after reading the source for sshclient it's existence doesn't make sense to me)

It should be possible to create forwarders to re-create the same effect that we already have with the ssh-forwarder today.

@guillaumerose I have been borrowing a lot of code from this repository to build a version of the virtual network that acts more like a configurable service and the API is written to be more consumable. When I finish the ipv6 support I'll be pushing it to github. If you're interested I can shoot you an email to see what you think. What I need it for is a cross-platform container VM solution that I'm building, but it should drive more than just my own hobby project.

Reply:
For IPv6, I have this PR opened with everything working (https://github.com/containers/gvisor-tap-vsock/pull/59). I don't merge it because I don't like one hack I had to write :) 

> If you're interested I can shoot you an email to see what you think.

Yes, good idea! Thanks 👍

Reply:
A single SSH connection is used and it will reconnect if EOF error is received from a read of 0 bytes on the underlying tcp connection.

Also added tests for unix2unix and unix2tcp forwarders. 

@guillaumerose tool I'm working on pretty much just borrows your IPv6 WIP code. It looked mostly fine to me, aside from some hard coded instances of "fe80::1" in places. Once I have more work done on DHCP for v6, that PR is on my giveback list this year.

Reply:
Thanks for the new commits. A part from the lock, which I am not sure, it looks good. 

@n1hility has more experience with ssh connections and keepalive. Can you take a look @n1hility ?

Reply:
The sshClient connection has been moved into it's own anonymous function so the locking would only apply to the relevant tasks. Being surgical with the lock in any other way seems like it would get racy. The lock shouldn't be necessary for sshClient.Dial.

Reply:
Sorry just had a chance to take a look I agree @protosam that the nested stream dial should be safe since the ssh client does have a locking model around internal message handling. Although regrading the heartbeat point @guillaumerose brings up, I do think you want reconnection logic here since long running ssh connections can easily timeout from either ssh, networking, netfilter policy, or cycling sshd. This is where locking becomes important for this type of dispatch model since you need a happens-before edge around changes to the sshClient handle. The role of the bastion code in sshclient is to wrap the connection over multiple reconnects and handle changes .  If dialing fails you can test the connection by sending a heartbeat (see the code in ssh_forwarder to see what I mean), and then tear down and rebuild the connection. Reestablishing (and the first establishment) might be rocky, so I recommend doing a similar backoff retry loop as well. One other thing you might want to look into when it comes to passing the dialed unix socket back to gvisor, is if the gvisor proxying logic manages half-close. The nested ssh streaming relies on a CloseWrite to gracefully terminate the client. I suspect gvisor does do this, but might be worth checking since otherwise you can get abrupt or delayed closes on interactive sessions. I have a local branch with the windows related changes I was referring to earlier, so when I finish that off tomorrow I can double check the gvisor thing if you dont get to it first.